kotest-assertions-arrow-fx-coroutines/src/commonMain/kotlin/io/kotest/assertions/arrow/fx/coroutines/ResourceExtension.kt

```diff
@@ -0,0 +1,104 @@
+package io.kotest.assertions.arrow.fx.coroutines
+
+import arrow.fx.coroutines.Resource
+import io.kotest.core.extensions.MountableExtension
+import io.kotest.core.listeners.AfterSpecListener
+import io.kotest.core.listeners.TestListener
+import io.kotest.core.spec.Spec
+import io.kotest.core.test.TestCase
+import io.kotest.core.test.TestResult
+import io.kotest.core.test.TestType
+import io.kotest.core.test.isRootTest
+
+public enum class LifecycleMode {
+  Spec, EveryTest, Leaf, Root
+}
+
+public fun <A> Resource<A>.extension(lifecycleMode: LifecycleMode = LifecycleMode.Spec): ResourceExtension<A> =
+  ResourceExtension(this, lifecycleMode)
+
+/**
+ * Ensures that all finalizers of [resource] are invoked for the specified [LifecycleMode].
+ * Installing this [MountableExtension] returns a [TestResource]
+ * from which you can extract initialized [A] using [TestResource.invoke] or [TestResource.get].
+ *
+ * ```kotlin
+ * class Connection(val name: String) {
+ *   fun open() = println("Open $name")
+ *   fun compute() = "Result of $name"
+ *   fun release(ex: ExitCase) = println("Closing $name with $ex")
+ * }
+ *
+ * fun connection(name: String): Resource<Connection> =
+ *   Resource(
+ *     { Resource(name) },
+ *     { conn, ex -> conn.release(ex) }
+ *   )
+ *
+ * class ResourceSpec : StringSpec({
+ *   val conn: TestResource<Connection> =
+ *     install(ResourceExtension(connection("DB")))
+ *
+ *   "invoke" {
+ *     conn().compute() shouldBe "Result of DB"
+ *   }
+ *
+ *   "get" {
+ *     conn.get().compute() shouldBe "Result of DB"
+ *   }
+ *
+ *   afterSpec {
+ *     shouldThrow<TestResource.AlreadyClosedException> {
+ *       conn.get()
+ *     }
+ *   }
+ * })
+ * ```
+ *
+ * `conn` cannot be used during or after `afterSpec` since at that point the resource is already released,
+ * and it will result in [TestResource.AlreadyClosedException]
+ *
+ * @param lifecycleMode allows installing the [Resource] for other scopes besides the default [LifecycleMode.Spec].
+ */
+public class ResourceExtension<A>(
+  public val resource: Resource<A>, public val lifecycleMode: LifecycleMode = LifecycleMode.Spec
+) : MountableExtension<A, TestResource<A>>, TestListener, AfterSpecListener {
```

Comment:
Sorry if the following questions are stupid, I don't know enough Arrow.. 😅 

Is the user (of arrow + kotest) most interested in interacting with `A` or `TestResource<A>` after installing the extension? The included tests only interacts with `A` so perhaps it would be better to have `ResourceExtension<A, A> : MountableExtension<A, A>` which doesn't even expose the resource? It seems to me that the `TestResource` will always be acquired in the scope of the test anyhow, so is there any reason a user might want to interact with it?

Reply:
Reason I'm asking is because I believe the pattern:

```kotlin
val sub = ResourceUnderTest() 

install(sub.asResource().extension()) { configure() }
```

would look normally look like:


```kolin
val sub = install(ResourceUnderTest().asResource().extension()) {
  configure()
}
```
But since the `ResourceExtension` materializes a `TestResource` you can't do that.

You also can't do this, since init-block is non-suspending
```kolin
val sub = install(ResourceUnderTest().asResource().extension()) {
  configure()
}.get() // or just invoke
```

Reply:
☝🏼  take this with a pinch of salt. I'm not extremely used to `MountableExtension`s yet so I might be talking out of my ass, but I think the questions are relevant

Reply:
Hey @Kantis,

Thanks for the review and remarks!

The user will always want to interact with `A` directly, the problem is that `install` does not support `suspend` which means it's only _safe_ to access from within `TestScope`.

I tried to model it in such a way that it's impossible to access it from a `non-suspend` context, this would then be within the `init` block. Otherwise we'd have to implicitly use `runBlocking` under the hood to initialise the `Resource`, which are in a lot of cases doing blocking I/O, etc.

IIRC `ResourceExtension<A, A> : MountableExtension<A, A>` would not have this same guarantees in terms of preventing the users to access the `A` from outside of its initialised state.

The tests here are not a good representation of how it would actually be used in reality, since to tests this I have to expose an "unsafe resource" so we can verify the interactions/state. Normally domain code would expose something like `Resource<Database>` and then the user would install, and use it like so:

```kotlin
fun database(config: HikariConfig): Resource<Database> = resource {
  val hikari = Resource.fromCloseable { HikariDataSource(config) }.bind()
  Resource({ Database.connect(hikari) }, { db, _ -> db.close() }).bind()
}

class MySpec : StringSpec({
  val testDb = install(database(config).extension()) {
    configure()
  }
  
  // testDb.get() not allowed requires supension
  // runBlocking { testDb.get() } forcefully initialise anyway, while still respecting the `LifecycleMode`.
  
  beforeSpec {
    testDb.get() // allowed since this is supports suspension
  }
  
  afterTest {
    testDb.get() // Also fine since it also supports suspension
  }
  
  "test" {
    testDb().shouldNotBeNull()
  }
})
```